software_checklist.md

```diff
@@ -0,0 +1,141 @@
+
+This is your overall judgement on the level of confidence based on all the aspects of the checklist. There is no formulaic way to arrive at this overall assessment based on the individual checklist answers but please explain how the measures in place combine to reach this level of confidence and make clear any caveats (eg applies for certain ways of using the software and not others).
+
+> The code is well crafted and organised, with documentation, and automated unit and regression testing via continuous integration. Execution environment is specified using [conda](https://docs.conda.io/en/latest/). It is planned that data integrity and provenance for reproducible model runs will be assured by use of the SCRC data pipeline, which is not yet fully available. Further work will be required to integrate with this. There are some areas where important improvements can are still being made (see below and the [project board](https://github.com/orgs/ScottishCovidResponse/projects/6)).
+
+## Checklist
+
+Please use a statement from this list: "Sufficiently addressed", "Some work remaining or caveats", or "Needs to be addressed" to begin each response.
+
+Additionally, for each question please explain the situation and include any relevant links (eg tool dashboards, documentation). The sub bullet points are to make the scope of the question clear and should be covered if relevant but do not have to be answered individually.
+
+### Can a run be repeated and reproduce exactly the same results?
+
+- How is stochasticity handled?
+- Is sufficient meta-data logged to enable a run to be reproduced: Is the exact code version recorded (and whether the repository was "clean"), including versions of dependent libraries (e.g. an environment.yml file or similar) along with all command line arguments and the content of any configuration files?
+- Is there up-to-date documentation which explains precisely how to run the code to reproduce existing results?
+
+> - [x] Sufficiently addressed
+> - [ ] Some work remaining or caveats
+> - [ ] Needs to be addressed
+>
+> Stochasticity is handled by using a seed from the data pipeline to generate a “random state” object which is passed to functions that generate random numbers. Where multiple processes are used, seeds for each process / iteration are generated based on a single initial seed, these are assured by `numpy` to be independently random. Git url and sha are saved when the software is run. Files that specify the `conda` environment are present: [spec-file.txt](spec-file.txt) and [environment.yml](environment.yml). Documentation explaining how to run the software with some test data can be found in the [readme](readme.md). We have not yet produced "real" results based on provenanced, reliable source data - this work is ongoing and reproducibility will be assured by a combination of this software and the SCRC data pipeline.
```

Comment:
@aflag and @FiodorG - have I described the random number generating OK?

Reply:
Maybe worth mentioning as well, the initial seed used to generate child seeds for each process is ran through a numpy SeedSequence, which "mixes sources of entropy in a reproducible way to set the initial state for independent and very probably non-overlapping BitGenerators".

As in generating different seeds in such a way gives some guarantees that random numbers generated for every separate run will be independent with high probability. Which wouldn't be the case if we simply used manually inputted seed 10, 11, 12, 13, etc to each process.

Likewise, @aflag mentioned something about using an entropy based initial seed which provides better results which we use.

Reply:
Thanks @FiodorG - I've amended this.